---
title: "DEAGO Analysis Report"
date: "`r format(Sys.time(), '%D')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 3  
    number_sections: true
    theme: paper
---

<style>
table {
  white-space: nowrap;
}
</style>

# Introduction

This report was generated by the WTSI Pathogen Informatics Differential Expression Pipeline. It should only be used as an overview of your RNA-Seq data and is not intended to be a final analysis.

The R package [DEAGO](https://github.com/vaofford/deago) is a wrapper which contains functions to generate QC plots, perform differential expression analysis with [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html) and GO term enrichment analysis with [topGO](http://bioconductor.org/packages/release/bioc/html/topGO.html).

The [Pipeline configuration](#Pipeline configuration) section below gives the details of the files and settings used to generate this report.  These have been imported from deago_go_keep.config.  Your results can be found in **/Users/vo1/sanger-pathogens/perl/Bio-Deago**.

```{r setup_libraries, include=FALSE}
library(deago)
library(DT)
knitr::opts_chunk$set(echo=TRUE)
```

# Pipeline configuration 
```{r config, message=FALSE}
parameters <- importConfig("deago_go_keep.config")
resultsDir <- makeResultDir(parameters$results_directory, parameters$keep_images)
parameters$results <- resultsDir
```

```{r printConfig, echo=FALSE}
datatable(t(as.data.frame(parameters)), options = list(dom = 't', pageLength = 15), colnames=c('parameter', 'value'))
```

# Imported data summary

The summary table below contains the total number of differentially expressed genes and the number of up-regulated (lfc > 2) and down-regulated (lfc < -2) genes for each contrast (padj < `r ifelse(is.na( parameters$alpha ),parameters$alpha,0.05)`).

```{r targets}
targets <- importTargets(parameters$targets_file)
countdata <- readCountData(targets, parameters$counts_directory, parameters$gene_ids, data_column=5, skip=0, sep=',')
```

```{r printTargets, echo=FALSE}
datatable(targets, rownames=FALSE, datatable(targets, rownames=FALSE, options = list(pageLength = 10, autoWidth = TRUE, scrollX = TRUE, scrollCollapse = TRUE, columnDefs = list(list(className = 'dt-center', targets = 0:ncol(targets)-1)))))
```


# DESeq2 analysis
```{r analysis, message=FALSE}
dds <- runDESeqAnalysis(targets, countdata, parameters)
```

```{r annotation}
if ("annotation_file" %in% names(parameters)) {
  dds <- annotateDataset(dds, parameters)
}
```

# QC plots
## Reads per sample
This plot shows the sum of gene read counts for each sample. Bar colours represent experimental condition.
```{r countPlot, fig.width=9, fig.height=7}
plotReadCounts(dds, resultsDir)
```

## Sample-to-sample distances
This plot gives an overview of how the samples cluster based on their euclidean distance using the regularized log transformed count data. It is a complementary figure to the PCA plot.
```{r plotSampleDistances, results="hide", fig.keep="all", warnings = FALSE, fig.width=9, fig.height=7}
plotSampleDistances(dds, resultsDir)
```

##PCA plot
This plot shows the first two principal components which explain the variability in the data using the regularized log count data.
```{r pcaPlot, warning=FALSE, fig.width=9, fig.height=7}
pcaPlot(dds,resultsDir)
```

##Cook's distances
Cook's distance is a measure of how mucha single sample is influencing the fitted coefficients for a gene, and a large value of Cook's distance is intended to indicate an outlier count. This plot can be used to see if one sample has much higher Cook's distances than the other samples.
```{r plotCooks, results="hide", fig.keep="all", warning=FALSE, comment=FALSE, fig.width=9, fig.height=7}
plotCooks(dds,resultsDir)
```

##Density plot
Density plot showing the distribution of normalised read counts per sample.
```{r plotDensity, results="hide", fig.keep="all", comment=FALSE, warning = FALSE, fig.width=9, fig.height=7}
plotDensity(dds,resultsDir)
```

##Dispersion plot
This dispersion estimate plot shows the gene-wise estimates (black), the fitted values (red), and the final maximum a posteriori estimates used in testing (blue).
```{r plotDispersionEstimates, results="hide", fig.keep="all", warning = FALSE, fig.width=9, fig.height=7}
plotDispersionEstimates(dds, resultsDir)
```

#Pairwise contrasts

The summary table below contains the total number of differentially expressed genes and the number of up-regulated (lfc > 2) and down-regulated (lfc < -2) genes for each contrast (padj < `r ifelse(is.na( parameters$alpha ),parameters$alpha,0.05)`).

```{r contrasts, echo=TRUE}
contrasts <- getContrasts(dds, parameters)
writeContrasts(dds, contrasts, resultsDir)
```

```{r contrast_summary, echo=TRUE}
contrast_summary <- contrastSummary(contrasts, parameters)
datatable(contrast_summary, options = list(dom = 't', colnames=c('contrast', 'up-regulated','down-regulated','total'), columnDefs = list(list(className = 'dt-center', targets = 1:ncol(contrast_summary)))))
```

## Venn diagram
The Venn diagram below represents the number of differentially expressed genes (padj < `r ifelse(is.na( parameters$alpha ),parameters$alpha,0.05)` that are up-regulated (lfc > 2) or down-regulated (lfc < -2) in each contrast. 
```{r venn, echo=TRUE, fig.width=7,fig.height=7}
venn_counts <- getVennCounts(contrasts)
plotVenn(venn_counts, resultsDir)
```

## 0b_vs_0a

#### MA plot
The MA plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored red if the adjusted p-value is less than 0.05. Points which fall out of the window are plotted as open triangles pointing either up or down.
```{r 0b_vs_0a.MA, results='asis'}
plotContrastMA(contrasts$'0b_vs_0a', resultsDir, geneLabels=TRUE)
```

#### Volcano plot
The volcano plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored where the p-value is less than 0.05. Green points represent down-regulated genes (log2FoldChange < -2) and orange points represent up-regulated genes (log2FoldChange > 2).
```{r 0b_vs_0a.volcano, results='asis'}
plotVolcano(contrasts$'0b_vs_0a', resultsDir, geneLabels=TRUE)
```

### Differential expression
The table below contains differntially expressed genes (adjusted p-value < 0.01, log2 fold change <= 2 | >= 2)
```{r 0b_vs_0a.DEA, results='asis'}
prepareContrastTable(contrasts$'0b_vs_0a')
```

### GO term enrichment
```{r 0b_vs_0a.GOanalysis, echo=TRUE, warning=FALSE, comment=FALSE}
go_tables <- suppressMessages(runGOanalysis(dds, contrasts, parameters))
writeGOtables(go_tables, resultsDir)
```

#### GO term enrichment - BP
The table below contains a list of the top 20 signicant GO terms identified by the weight01 method and their associated p-values.
```{r 0b_vs_0a.BP, message=FALSE, comment=FALSE, warning=FALSE, results='asis'}
prepareGOtable(go_tables$'0b_vs_0a_BP')
```




#### GO term enrichment - MF
The table below contains a list of the top 20 signicant GO terms identified by the weight01 method and their associated p-values.
```{r 0b_vs_0a.MF, message=FALSE, comment=FALSE, warning=FALSE, results='asis'}
prepareGOtable(go_tables$'0b_vs_0a_MF')
```




## 0c_vs_0a

#### MA plot
The MA plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored red if the adjusted p-value is less than 0.05. Points which fall out of the window are plotted as open triangles pointing either up or down.
```{r 0c_vs_0a.MA, results='asis'}
plotContrastMA(contrasts$'0c_vs_0a', resultsDir, geneLabels=TRUE)
```

#### Volcano plot
The volcano plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored where the p-value is less than 0.05. Green points represent down-regulated genes (log2FoldChange < -2) and orange points represent up-regulated genes (log2FoldChange > 2).
```{r 0c_vs_0a.volcano, results='asis'}
plotVolcano(contrasts$'0c_vs_0a', resultsDir, geneLabels=TRUE)
```

### Differential expression
The table below contains differntially expressed genes (adjusted p-value < 0.01, log2 fold change <= 2 | >= 2)
```{r 0c_vs_0a.DEA, results='asis'}
prepareContrastTable(contrasts$'0c_vs_0a')
```

### GO term enrichment
```{r 0c_vs_0a.GOanalysis, echo=TRUE, warning=FALSE, comment=FALSE}
go_tables <- suppressMessages(runGOanalysis(dds, contrasts, parameters))
writeGOtables(go_tables, resultsDir)
```

#### GO term enrichment - BP
The table below contains a list of the top 20 signicant GO terms identified by the weight01 method and their associated p-values.
```{r 0c_vs_0a.BP, message=FALSE, comment=FALSE, warning=FALSE, results='asis'}
prepareGOtable(go_tables$'0c_vs_0a_BP')
```




#### GO term enrichment - MF
The table below contains a list of the top 20 signicant GO terms identified by the weight01 method and their associated p-values.
```{r 0c_vs_0a.MF, message=FALSE, comment=FALSE, warning=FALSE, results='asis'}
prepareGOtable(go_tables$'0c_vs_0a_MF')
```




## 0c_vs_0b

#### MA plot
The MA plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored red if the adjusted p-value is less than 0.05. Points which fall out of the window are plotted as open triangles pointing either up or down.
```{r 0c_vs_0b.MA, results='asis'}
plotContrastMA(contrasts$'0c_vs_0b', resultsDir, geneLabels=TRUE)
```

#### Volcano plot
The volcano plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored where the p-value is less than 0.05. Green points represent down-regulated genes (log2FoldChange < -2) and orange points represent up-regulated genes (log2FoldChange > 2).
```{r 0c_vs_0b.volcano, results='asis'}
plotVolcano(contrasts$'0c_vs_0b', resultsDir, geneLabels=TRUE)
```

### Differential expression
The table below contains differntially expressed genes (adjusted p-value < 0.01, log2 fold change <= 2 | >= 2)
```{r 0c_vs_0b.DEA, results='asis'}
prepareContrastTable(contrasts$'0c_vs_0b')
```

### GO term enrichment
```{r 0c_vs_0b.GOanalysis, echo=TRUE, warning=FALSE, comment=FALSE}
go_tables <- suppressMessages(runGOanalysis(dds, contrasts, parameters))
writeGOtables(go_tables, resultsDir)
```

#### GO term enrichment - BP
The table below contains a list of the top 20 signicant GO terms identified by the weight01 method and their associated p-values.
```{r 0c_vs_0b.BP, message=FALSE, comment=FALSE, warning=FALSE, results='asis'}
prepareGOtable(go_tables$'0c_vs_0b_BP')
```




#### GO term enrichment - MF
The table below contains a list of the top 20 signicant GO terms identified by the weight01 method and their associated p-values.
```{r 0c_vs_0b.MF, message=FALSE, comment=FALSE, warning=FALSE, results='asis'}
prepareGOtable(go_tables$'0c_vs_0b_MF')
```



