---
title: "DEAGO Analysis Report"
date: "`r format(Sys.time(), '%D')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 3  
    number_sections: true
    theme: paper
---

<style>
table {
  white-space: nowrap;
}
</style>

# Introduction

This report was generated by the WTSI Pathogen Informatics Differential Expression And GO enrichment analysis pipeline. It should only be used as an overview of your RNA-Seq data and is not intended to be a final analysis.

The R package [DEAGO](https://github.com/vaofford/deago) is a wrapper which contains functions to generate QC plots, perform differential expression analysis with [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html) and GO term enrichment analysis with [topGO](http://bioconductor.org/packages/release/bioc/html/topGO.html).

The [Pipeline configuration](#Pipeline configuration) section below gives the details of the files and settings used to generate this report.  These have been imported from /lustre/scratch118/infgen/pathdev/vo1/data/lrt_analysis/deago.config.

```{r setup_libraries, include=FALSE}
library(deago)
library(DT)
library(DESeq2)
knitr::opts_chunk$set(echo=TRUE)
```

# Pipeline configuration 
```{r config, message=FALSE}
parameters <- importConfig("/lustre/scratch118/infgen/pathdev/vo1/data/lrt_analysis/deago.config")
resultsDir <- makeResultDir(parameters$results_directory, parameters$keep_images)
parameters[['results']] <- resultsDir
```

```{r printConfig, echo=FALSE}
datatable(t(as.data.frame(parameters)), options = list(dom = 't', pageLength = 20), colnames=c('parameter', 'value'))
```

# Imported data summary

The summary table below contains the total number of differentially expressed genes and the number of up-regulated (lfc > 2) and down-regulated (lfc < -2) genes for each contrast (padj < 10E-12).

```{r targets}
targets <- importTargets(parameters$targets_file)
countdata <- readCountData(targets, parameters$counts_directory, parameters$gene_ids, data_column=7, skip=1, sep='\t')
```

```{r printTargets, echo=FALSE}
datatable(targets, rownames=FALSE, options = list(pageLength = 10, scrollX = TRUE, scrollCollapse = TRUE, columnDefs = list(list(className = 'dt-center', targets = 0:ncol(targets)-1))))
```


# DESeq2 analysis
```{r analysis, message=FALSE}
coldata <- DataFrame( cell_type=factor( tolower(targets$cell_type) ),
                      treatment=factor( tolower(targets$treatment) ),
                      condition=factor( tolower(targets$condition) ) )

dds <- DESeqDataSetFromMatrix(countData=countdata, colData=coldata, design=~cell_type+treatment+cell_type:treatment)

dds$cell_type <- relevel(dds$cell_type, ref = "ko")
dds$treatment <- relevel(dds$treatment, ref = "ctrl")

dds <- DESeq(dds, test="LRT", reduced=~cell_type+treatment)
dds <- annotateDataset(dds, parameters)
```

```{r annotation}
if ("annotation_file" %in% names(parameters)) {
  dds <- annotateDataset(dds, parameters)
}
```

# QC plots
## Total read counts per sample
This reads per sample plot shows the total read counts (raw and normalized) for each sample. Bar colours represent experimental condition.
```{r readCountPlot, fig.width=9, fig.height=11}
plotReadCounts(dds, resultsDir)
```

## Null count percentage per sample
This null count percentage per sample plot shows the percentage of genes which have no counts (raw and normalized) for each sample. Bar colours represent experimental condition.
```{r nullCountPlot, fig.width=9, fig.height=11}
plotNullCounts(dds, resultsDir)
```

## Sample-to-sample distances
The sample-to-sample distance plot gives an overview of how the samples cluster based on their euclidean distance using the regularized log transformed count data.
```{r plotSampleDistances, results="hide", fig.keep="all", warnings = FALSE, fig.width=12, fig.height=11}
plotSampleDistances(dds, resultsDir)
```

##Principal component analysis (PCA)
### PCA plot
The Principal Component Analysis (PCA) plot shows the first two principal components which explain the variability in the data using the regularized log count data.  
```{r pcaPlot, warning=FALSE, fig.width=12, fig.height=11}
pc_list <- getPrincipalComponents(dds)
pcaPlot(pc_list,resultsDir)
```

### PC scree plot
The principal components (PC) scree plot shows the percentage contribution of each PC. 
```{r pcaScreePlot, warning=FALSE, fig.width=9, fig.height=7}
pcaScreePlot(pc_list,resultsDir)
```

### PC summary
The PC summary shows the percentage contribution (bars in scree plot) and cumulative total (points in scree plot) for each PC.
```{r pcaSummary, warning=FALSE}
pcaSummary(pc_list)
```

##Cook's distances
Cook's distance is a measure of how much a single sample is influencing the fitted coefficients for a gene. A large value of Cook's distance is intended to indicate an outlier count. 
```{r plotCooks, results="hide", fig.keep="all", warning=FALSE, comment=FALSE, fig.width=9, fig.height=7}
plotCooks(dds,resultsDir)
```

##Density plot
The density plot shows the distribution of normalised read counts per sample.
```{r plotDensity, results="hide", fig.keep="all", comment=FALSE, warning = FALSE, fig.width=11, fig.height=12}
plotDensity(dds,resultsDir)
```

##Dispersion plot
The dispersion estimate plot shows the gene-wise estimates (black), fitted values (red) and final maximum a posteriori estimates used in testing (blue).
```{r plotDispersionEstimates, results="hide", fig.keep="all", warning = FALSE, fig.width=9, fig.height=7}
plotDispersionEstimates(dds, resultsDir)
```

#Pairwise contrasts

```{r contrasts, echo=TRUE}
lrt_results <- results(dds, alpha=as.numeric(parameters$qvalue))
lrt_results$symbol <- mcols(dds)$symbol

contrasts <- list('lrt'=lrt_results)
writeContrasts(dds, contrasts, resultsDir)
```

## Contrast summary

The summary table below contains the total number of differentially expressed genes and the number of up-regulated (lfc > 2) and down-regulated (lfc < -2) genes for each contrast (adjusted p-value < 10E-12).

```{r contrast_summary, echo=TRUE}
contrast_summary <- contrastSummary(contrasts, parameters)
datatable(contrast_summary, options = list(dom = 't', colnames=c('contrast', 'up-regulated','down-regulated','total'), columnDefs = list(list(className = 'dt-center', targets = 1:ncol(contrast_summary)))))
```

#### MA plot
The MA plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored red if the adjusted p-value is less than 0.05. Points which fall out of the window are plotted as open triangles pointing either up or down.
```{r lrt.MA, results='asis'}
plotContrastMA(contrasts$'lrt', resultsDir, geneLabels=TRUE)
```

#### Volcano plot
The volcano plot below compares the mean of the normalized counts against the log fold change, showing one point per gene.  Points will be colored where the adjusted p-value is less than 0.05. Green points represent down-regulated genes (lfc < -2) and orange points represent up-regulated genes (lfc > 2).
```{r lrt.volcano, results='asis'}
plotVolcano(contrasts$'lrt', resultsDir, geneLabels=TRUE)
```

### Differential expression
The table below contains differntially expressed genes (adjusted p-value < 0.01, log2 fold change <= 2 | >= 2).
```{r lrt.DEA, results='asis'}
prepareContrastTable(contrasts$'lrt')
```

# R session
```{r session, echo=FALSE}
sessionInfo()
```
